ext/fileinfo/fileinfo.c

```diff
@@ -275,181 +275,198 @@ PHP_FUNCTION(finfo_set_flags)
-			php_stream_statbuf ssb;
-
-			// Implementation is used for both finfo_file() and mimetype_emu()
-			int buffer_param_num = (mimetype_emu ? 1 : 2);
-			if (buffer == NULL || buffer_len == 0) {
-				zend_argument_must_not_be_empty_error(buffer_param_num);
-				goto clean;
-			}
-			if (CHECK_NULL_PATH(buffer, buffer_len)) {
-				zend_argument_type_error(buffer_param_num, "must not contain any null bytes");
-				goto clean;
-			}
+	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "OP|lr!", &self, finfo_class_entry, &path, &options, &zcontext) == FAILURE) {
+		RETURN_THROWS();
+	}
+	FILEINFO_FROM_OBJECT(finfo, self);
+	magic = finfo->magic;
 
-			wrap = php_stream_locate_url_wrapper(buffer, &tmp2, 0);
+	if (UNEXPECTED(ZSTR_LEN(path) == 0)) {
+		zend_argument_must_not_be_empty_error(2);
+		RETURN_THROWS();
+	}
+	php_stream_context *context = php_stream_context_from_zval(zcontext, false);
 
-			if (wrap) {
-				php_stream *stream;
-				php_stream_context *context = php_stream_context_from_zval(zcontext, 0);
+	/* Set options for the current file/buffer. */
+	if (options) {
+		magic_setflags(magic, options);
+	}
 
-#ifdef PHP_WIN32
-				if (php_stream_stat_path_ex(buffer, 0, &ssb, context) == SUCCESS) {
-					if (ssb.sb.st_mode & S_IFDIR) {
-						ret_val = mime_directory;
-						goto common;
-					}
-				}
-#endif
+	const char *ret_val = php_fileinfo_from_path(magic, path, context);
+	if (ret_val) {
+		RETVAL_STRING(ret_val);
+	} else {
+		RETVAL_FALSE;
+	}
 
-				stream = php_stream_open_wrapper_ex(buffer, "rb", REPORT_ERRORS, NULL, context);
+	/* Restore options */
+	if (options) {
+		magic_setflags(magic, options);
+	}
+}
 
-				if (!stream) {
-					RETVAL_FALSE;
-					goto clean;
-				}
+/* {{{ Return information about a string buffer. */
+PHP_FUNCTION(finfo_buffer)
+{
+	zval *self;
+	struct magic_set *magic = NULL;
+	zend_string *buffer = NULL;
+	zend_long options = 0;
+	zval *dummy_context = NULL;
+	php_fileinfo *finfo = NULL;
 
-				if (php_stream_stat(stream, &ssb) == SUCCESS) {
-					if (ssb.sb.st_mode & S_IFDIR) {
-						ret_val = mime_directory;
-					} else {
-						ret_val = (char *)magic_stream(magic, stream);
-					}
-				}
+	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "OS|lr!", &self, finfo_class_entry, &buffer, &options, &dummy_context) == FAILURE) {
+		RETURN_THROWS();
+	}
+	FILEINFO_FROM_OBJECT(finfo, self);
+	magic = finfo->magic;
 
-				php_stream_close(stream);
-			}
-			break;
-		}
-		EMPTY_SWITCH_DEFAULT_CASE()
+	/* Set options for the current file/buffer. */
+	if (options) {
+		magic_setflags(magic, options);
 	}
 
-common:
+	const char *ret_val = magic_buffer(magic, ZSTR_VAL(buffer), ZSTR_LEN(buffer));
+
 	if (ret_val) {
 		RETVAL_STRING(ret_val);
 	} else {
 		php_error_docref(NULL, E_WARNING, "Failed identify data %d:%s", magic_errno(magic), magic_error(magic));
 		RETVAL_FALSE;
 	}
 
-clean:
-	if (mimetype_emu) {
-		magic_close(magic);
-	}
-
 	/* Restore options */
 	if (options) {
```

Comment:
I'd move this before the if so that the return values can be set via RETURN_ instead of RETVAL_